### ISSUE TITLE
Terminal font color with minimum contrast ratio

### ISSUE THREAD

**Issue by jbericat** [gnome-terminator/terminator#894]
Hello, 

I'm trying to find a way to enable a "minimum contrast ratio" in Terminator in the same way I can do it in vscode terminal (see https://code.visualstudio.com/docs/terminal/appearance#_minimum-contrast-ratio). For example, my Terminator terminal looks like this:

![image](https://github.com/gnome-terminator/terminator/assets/8354056/4296db49-e381-4839-aa9c-d768ad534e11)

But I want to make it change the font color dynamically when a background color does not provide enough contrast, like in vscode terminal:

![image](https://github.com/gnome-terminator/terminator/assets/8354056/c2eea0ef-2db1-44d6-8083-0d24106c50d8)

Maybe there is a feature to enable this already, but I'm unable to find it -so I decided to come here to ask.

Thanks!

**Comment by rcalixte** [comment]
> I'm trying to find a way to enable a "minimum contrast ratio" in Terminator in the same way I can do it in vscode

VSCode uses the [Xterm.js](https://github.com/xtermjs/xterm.js) library to accomplish the contrast ratio functionality. In theory, the same math could be applied to create a contrast function/option here. I think it could help with the usability of the default themes. If this feature is determined to be desired by the project's maintainers, I wouldn't mind giving it a first pass.

**Comment by egmontkob** [comment]
This needs support from VTE; see https://gitlab.gnome.org/GNOME/vte/-/issues/2248 for vague ideas (in brainstorming phase) about this possible feature.

**Comment by rcalixte** [comment]
> This needs support from VTE; see https://gitlab.gnome.org/GNOME/vte/-/issues/2248 for vague ideas (in brainstorming phase) about this possible feature.

For hard-coded prompts like the example, I think you're right. The remaining palette and color schemes are defined within Terminator though. From there, it would be a matter of calculating the acceptable range(s) for contrast purposes and then over-writing it upon display. For color schemes defined outside of those parameters (like the custom prompt), they probably fall out of scope but I would need to research more.

**Comment by egmontkob** [comment]
Terminator allows you to configure the main 18 colors (default fg and bg, plus 16 standard palette colors). Any of these can be used as foreground or as background and can be combined arbitrarily with each other to achieve up to 18*18 combinations (some of which are obviously unreadable), and then we haven't talked about the extended 256-color support or direct RGB colors.

A "minimum contrast ratio" feature doesn't mean to once modify these 18 colors into a fixed set of 18 new ones so that all 18*18 combinations of these new ones (except for the 18 where the fg and bg are literally the same) look good. The way a certain color is modified needs to depend on its context; to take the most trivial example: for each cell, the foreground color might need to be modified to be far away from the background color of that particular cell, whereas the background color should remain unchanged.

This definitely needs support from VTE which is not yet available.

**Comment by rcalixte** [comment]
> Terminator allows you to configure the main 18 colors (default fg and bg, plus 16 standard palette colors).

This is the implementation I'm referencing though. We only need to compare the palette colors to the fg and bg. I'm opting to go for the simplest route here since most folks probably have black or white terminal backgrounds but it should work for those cases in general. I agree that a broader "perfect" implementation would need better support upstream.

**Comment by egmontkob** [comment]
But then what you're wishing to have... does it differ in any way from the user wisely choosing a set of 18 colors? Disallowing colors like the bright cyan in the opening post (because white text is hard to read on it), forcing it to be a darker shade? Therefore forcing all the non-black non-white colors to be very much in the middle brightness, so that either black on white looks good on it?

OP's screenshots and request clearly demonstrate that white on bright cyan is hard to read, and he would prefer the bright cyan to remain the same and the foreground color to automatically change from white to black. This is definitely _not_ possible with current VTE. He can surely choose bright cyan to be somewhat darker, that doesn't need any support from Terminator that's not already there, namely the possibility of being able to pick the exact brightness of that color.

And then we have still only talked about either the fg or the bg being of the default fg or bg for each cell - something that is often not the case. There's no way you could come up with 16 or 18 colors so that all pairs look good.

And we still haven't solved the contrast issue with 256-color palette or direct RGB colors either.

No one stops you from choosing your preferred set of the 18 main colors, according to whatever criteria you have. This bugreport is not about that.

**Comment by rcalixte** [comment]
> And then we have still only talked about either the fg or the bg being of the default fg or bg for each cell - something that is often not the case. There's no way you could come up with 16 or 18 colors so that all pairs look good.


> And we still haven't solved the contrast issue with 256-color palette or direct RGB colors either.


That's exactly what [the library](https://github.com/xtermjs/xterm.js/blob/master/src/common/Color.ts) I mentioned before does. (To be clear, I would opt for a much simpler implementation.) At the end of the day, it's all hex values which means that we can calculate an appropriate distance (even a mirror value). They're all hard-coded when the application is launched or when any of the values get updated. So the only extra step is running a trivial calculation and storing the result in memory to be used in lieu of the exact palette string.







> No one stops you from choosing your preferred set of the 18 main colors, according to whatever criteria you have. This bugreport is not about that.

I recently switched to Terminator myself but it took me some time to find color palettes that worked for how I wanted it to. (And in the process, I think I found a few bugs but I haven't circled back yet. 😅) If I could've just hit the drop-down presets and enabled a contrast option, I wouldn't have had to spend this time. Even now, I see screenshots of colors clashing. If it's not an obscene amount of code and doesn't impose a maintenance burden, I think it's worth writing even if it might not address the initial issue perfectly.

**Comment by egmontkob** [comment]
> That's exactly what [the library](https://github.com/xtermjs/xterm.js/blob/master/src/common/Color.ts) I mentioned before does.

This is the relevant part of the source code of a particular terminal emulator which implements this feature. But that doesn't make it a _library_.

And it indeed implements the functionality we'd need to have. Given a fg and bg color of one particular cell, and a desired minimum contrast, it tells you how to adjust the fg color of that one cell. Not how to adjust the 18 global palette colors, but how to adjust the foreground of one cell, depending on the background of that very cell.

It is impossible for Terminator to implement this on top of current VTE. The feature has to be implemented in VTE, or at least a color mangling hook added to VTE which Terminator could hook up to and implement this feature.

**Comment by rcalixte** [comment]
> This is the relevant part of the source code of a particular terminal emulator which implements this feature. But that doesn't make it a _library_.

It's a single file within a library. 😅 


> And it indeed implements the functionality we'd need to have. Given a fg and bg color of one particular cell, and a desired minimum contrast, it tells you how to adjust the fg color of that one cell. Not how to adjust the 18 global palette colors, but how to adjust the foreground of one cell, depending on the background of that very cell.

My intention would be to implement that exact target functionality over an iterative array of the palette colors. The result would be stored in memory (so that the user's configuration is unaffected). Then those colors would be handed off to Terminator (or we could update real-time as well).

I'm not sure if we're talking about the same thing since we don't seem to be on the same page but effectively, the code I'm proposing to write would be to offer a configuration option that would create a pass-through filter of sorts from the configuration colors to Terminator's windows, defining ultimately how colors are shown to the user with the goal to prevent unsightly clashes and ease configuration and usability at the same time.

Think of it like this pseudocode:
```py
# ... somewhere in main()
if contrast_colors:
    new_bg, new_colors = contrast_colors(fg, bg, palette)

def contrast_colors(fg, bg, palette):
    new_palette = palette.copy()
    new_bg = bg

    for index, color in enumerate(palette):
        if tooCloseTo(color, bg):
            doHextMath(color, bg)
            new_palette[index] = result
        if tooCloseTo(fg, bg):
            result = doHexMath(fg, bg)
            new_bg = result

    return new_bg, new_palette

```

I hope this makes sense!

**Comment by egmontkob** [comment]
No, still absolutely does not make sense to me.

You can tell to VTE one single hex RGB code for the default foreground, one single hex RGB code for the default background, one single color for each of the 16 palette slots. And that color will be used by VTE for that purpose, regardless of context.

You tell VTE that your preferred shade of white is `#abcdef`. VTE, if asked to paint white on black, will use `#abcdef` as the foreground. If asked to paint white on cyan, it will use `#abcdef`. If asked to paint white on white, it will use `#abcdef`. If asked to use white as a background, it will use `#abcdef`.

What you would need to have is a context-dependant white. If used on black background then it remains white. If used on cyan background then it becomes black instead (as demonstrated in OP's screenshot) because otherwise the contrast isn't enough.

The color's exact value having to depend on its context (in particular: the foreground color having to depend on the background of that very same character cell) is something that is not possible with current VTE.

If you're good with a context-independent modification of colors then 1) you are absolutely clearly talking about something different than OP, and 2) nothing stops you from manually setting your desired 18 adjusted colors in Terminator right now.

I think I've explained it to you 5 times already (if I count as 1 the VTE issue I linked to). If you still don't get it, I see no way I could ever explain it to you. I'm stopping wasting my time on this conversation, I'm out.

**Comment by rcalixte** [comment]
> No, still absolutely does not make sense to me.

I would encourage you to read the source code here: [L788-L824](https://github.com/gnome-terminator/terminator/blob/master/terminatorlib/terminal.py#L788-L824)



> 2) nothing stops you from manually setting your desired 18 adjusted colors in Terminator right now.

Choosing a new background color today requires manually tuning it. I would rather spend time writing the code today to have a check box well into the future.



> I think I've explained it to you 5 times already (if I count as 1 the VTE issue I linked to). If you still don't get it, I see no way I could ever explain it to you. I'm stopping wasting my time on this conversation, I'm out.

I don't see the need for hostility. By all means, please see through the VTE issue to completion. There's no reason this can't be addressed on multiple fronts. Have a nice day!

**Comment by egmontkob** [comment]
> I would encourage you to read the source code here: [L788-L824](https://github.com/gnome-terminator/terminator/blob/master/terminatorlib/terminal.py#L788-L824)

Yes, this sets the fixed 18 or so colors. No context dependency. Absolutely clearly not what OP requested, and not what the code you linked to in xterm.js does.

After about 5 back and forth, you still don't understand the difference between altering the 18 "usual" colors once and using those altered versions globally everywhere, vs. altering the colors (even the extended 256 or 16 million ones) for each and every character cell individually for the needs of that very cell.

[Also: don't link to master in long-lived posts, the URL works now but will break as soon as the code is modified. Always link to a particular tag or commit id, unless in short-lived messages such as unarchived chat.]

> I don't see the need for hostility.

Hostility? What the heck is hostile in refusing to volunteer more time to explaining something that I've already failed to explain to you 5 times in a row? Surely someone could have explained more clearly, I'm sorry but it's not me.

> There's no reason this can't be addressed on multiple fronts.

Would love to see your Terminator patch addressing OP's request (with ~7 years of VTE development experience knowing that it's not possible :stuck_out_tongue:)

**Comment by rcalixte** [comment]
> No context dependency.

I said this in my initial reply. It's not a perfect solution for the customization presented here. It does work for the other instances of color clashes and that's why I think it's worth implementing. As a workaround, OP could take the computed resulting palette and use that. Again, not a perfect solution and probably not one that either of us would consider ideal but the original problem goes away and that is the goal I _am_ okay with.

> After about 5 back and forth, you still don't understand the difference between altering the 18 "usual" colors once and using those altered versions globally everywhere, vs. altering the colors (even the extended 256 or 16 million ones) for each and every character cell individually for the needs of that very cell.

In this case, you would be mistaken about what I understand (or don't) and what I'm intentionally paying mind to instead. I have no intention of implementing HDR for Terminator or anything arising to that level of complexity. In layman's terms, "I don't want light yellow fonts on a white background when I'm using presets." If there was a configuration option that would avoid that for me so I don't have to manually edit the palette, especially as a non-technical user, that would be valued.


> [Also: don't link to master in long-lived posts, the URL works now but will break as soon as the code is modified. Always link to a particular tag or commit id, unless in short-lived messages such as unarchived chat.]

I normally don't but this was rather spur of the moment. I've corrected the hyperlink.



> Would love to see your Terminator patch addressing OP's request (with ~7 years of VTE development experience knowing that it's not possible 😛)

I'm aware of your background. You see this as related to VTE. I am focused on what we can do within Terminator. I don't have a track record of long threads in Issues, mostly Pull Requests to resolve said Issues. There are options we can present here to work around some of the current limitations that exist within VTE. If it's a welcome feature, I'll be sure to cc you on the Pull Request. 😉
